### PR TITLE
[DPP-762][Self-service error codes] Automate generation of inventory of error categories.

### PR DESCRIPTION
**Pull request by pbatko-da**
Other changes:
- Fix ErrorCodeDocumentationGeneratorSpec not being in any bazel target.
- Fix inconsistent usage of deprecated, Deprecation and DeprecatedDocs annotation.

CHANGELOG_BEGIN
CHANGELOG_END

### CODE REVIEW

File: docs/source/app-dev/grpc/error-codes.rst [hunk 1]
```diff
@@ -88,201 +88,8 @@ The error categories allow to group errors such that application logic can be bu
 in a sensible way to automatically deal with errors and decide whether to retry
 a request or escalate to the operator.
 
-TransientServerFailure
```
Comment (pbatko-da) [inline docs/source/app-dev/grpc/error-codes.rst:91]: This is now generated automatically.

File: ledger/error/generator/app/src/main/scala/com/daml/error/generator/app/ErrorCategoryInventoryDocsGenApp.scala [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+package com.daml.error.generator.app
+
+import java.nio.file.{Files, Paths, StandardOpenOption}
+
+import com.daml.error.ErrorCategory
+import com.daml.error.generator.ErrorCodeDocumentationGenerator
+
+/** Generates error categories inventory as a reStructuredText
+  */
+object ErrorCategoryInventoryDocsGenApp {
```
Comment (pbatko-da) [inline ledger/error/generator/app/src/main/scala/com/daml/error/generator/app/ErrorCategoryInventoryDocsGenApp.scala:13]: Adapted from SelfServiceErrorCodes_MigrationGuideGen_App.scala

File: ledger/error/generator/lib/src/main/scala/com/daml/error/generator/ErrorCodeDocumentationGenerator.scala [hunk 3]
```diff
@@ -176,13 +171,57 @@ class ErrorCodeDocumentationGenerator(prefixes: Array[String] = Array("com.daml"
...
+  private val resolutionTypeName = classOf[Resolution].getTypeName.replace("$", ".")
+  private val descriptionTypeName = classOf[Description].getTypeName.replace("$", ".")
+  private val retryStrategyTypeName = classOf[RetryStrategy].getTypeName.replace("$", ".")
+
+  private val acceptedTypeNamesAnnotationsForErrorCodes =
+    Set(deprecatedTypeName, deprecatedDocsTypeName, explanationTypeName, resolutionTypeName)
+  private val acceptedTypeNamesAnnotationsForErrorCategories =
+    Set(descriptionTypeName, resolutionTypeName, retryStrategyTypeName)
+
+  def getErrorCategoryAnnotations(errorCategory: ErrorCategory): ErrorCategoryAnnotations = {
```
Comment (pbatko-da) [inline ledger/error/generator/lib/src/main/scala/com/daml/error/generator/ErrorCodeDocumentationGenerator.scala:199]: Adapted from SelfServiceErrorCodes_MigrationGuideGen_App.scala

File: ledger/error/generator/lib/src/test/scala/com/daml/error/generator/ErrorCodeDocumentationGeneratorSpec.scala [hunk 4]
```diff
@@ -1,74 +0,0 @@
...
-import com.daml.error.utils.testpackage.subpackage.MildErrors.NotSoSeriousError
-import com.daml.error.utils.testpackage.{DeprecatedError, SeriousError}
-import com.daml.error.{Deprecation, Explanation, Grouping, Resolution}
-import org.scalatest.flatspec.AnyFlatSpec
-import org.scalatest.matchers.should.Matchers
-
-import scala.annotation.nowarn
-
-@nowarn("msg=deprecated")
-class ErrorCodeDocumentationGeneratorSpec extends AnyFlatSpec with Matchers {
```
Comment (pbatko-da) [inline ledger/error/generator/lib/src/test/scala/com/daml/error/generator/ErrorCodeDocumentationGeneratorSpec.scala:16]: This class wasn't in any bazel target. 
Moved it to a different directory and fixed the test cases

File: ledger/error/src/main/scala/com/daml/error/ErrorCode.scala [hunk 5]
```diff
@@ -275,9 +275,8 @@ object ErrorCode {
 }
 
 // Use these annotations to add more information to the documentation for an error on the website
-case class Deprecation(deprecation: String) extends StaticAnnotation
 case class Explanation(explanation: String) extends StaticAnnotation
 case class Resolution(resolution: String) extends StaticAnnotation
 case class Description(description: String) extends StaticAnnotation
 case class RetryStrategy(retryStrategy: String) extends StaticAnnotation
-case class DeprecatedDocs(description: String) extends StaticAnnotation
+case class DeprecatedDocs(deprecation: String) extends StaticAnnotation
```
Comment (pbatko-da) [inline ledger/error/src/main/scala/com/daml/error/ErrorCode.scala:282]: `Deprecation` and `DeprecatedDocs` seemed to be redundant. Arbitrarily chose to use `DeprecatedDocs` and remove the other

File: ledger/error/src/test/utils/scala/com/daml/error/utils/testpackage/subpackage/MildErrors.scala [hunk 6]
```diff
@@ -1,41 +0,0 @@
-// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
-// SPDX-License-Identifier: Apache-2.0
-
-package com.daml.error.utils.testpackage.subpackage
-
-import com.daml.error._
-import com.daml.logging.LoggingContext
-
-@Explanation("Groups mild errors together")
-object MildErrors
```
Comment (pbatko-da) [inline ledger/error/src/test/utils/scala/com/daml/error/utils/testpackage/subpackage/MildErrors.scala:10]: Moved to MildErrorsParent.scala, where `MildErrors` is a nested within `MildErrorsParent`

File: ledger/error/generator/lib/src/test/scala/com/daml/error/generator/ErrorCodeDocumentationGeneratorSpec.scala (hunk 4 above)
Comment (pbatko-da) [reply ledger/error/generator/lib/src/test/scala/com/daml/error/generator/ErrorCodeDocumentationGeneratorSpec.scala:16]: fyi @fabiotudone-da @tudor-da

File: ledger/error/src/main/scala/com/daml/error/ErrorCode.scala (hunk 5 above)
Comment (tudor-da) [reply ledger/error/src/main/scala/com/daml/error/ErrorCode.scala:282]: Yes, they are redundant. This quirk stemmed from the impossibility of extracting the message from the Scala `@deprecated` annotation for both 2.12 and 2.13 using the same code ❗ Now that we've dropped 2.12, I would even revert to Fabio's change: Remove `DeprecatedDocs` and just use `deprecated`

**Review by tudor-da — APPROVED** [review]

**Pull request merged**